setup.cfg

```diff
@@ -45,3 +40,141 @@ ignore = E203,E501
 # D213 Multi-line docstring summary should start at the second line (reason: pep257 default)
 ignore = D107, D203, D213
 ignore_decorators = overrides
+
+[tox:tox]
+min_version = 4.0
+env_list =
+    # Parametrized environments.
+    # First parameter allows us to choose Python 3.8 or 3.10.
+    # Second parameter chooses how to define the environment:
+    #  withreq: using requirements-devel.txt
+    #  noreq: Without either requirements file (but including dev requirements)
+    # Third parameter selects the current unit tests (unit) or the legacy unit tests (legacy)
+    py{38,310}-{withreq,noreq}-{unit,legacy}
+skip_missing_interpreters = true
+labels =
+    # Minimal testing environments. run with `tox run-parallel -m test`
+    test = py38-withreq-{unit,legacy}
+    # Test in Python 3.10 from an empty environment.
+    future-test = py310-noreq-{unit,legacy}
+    # Environments to run in CI
+    ci = py{38,310}-withreq-{unit,legacy},py{38,310}-noreq-unit
+    # Just run the regular unit tests, not the legacy ones
+    unit = py{38,310}-{withreq,noreq}-unit
+    # Legacy unit tests only
+    legacy = py{38,310}-{withreq,noreq}-legacy
+set_env =
+    SNAPCRAFT_IGNORE_YAML_BINDINGS: "1"
+
+[testenv]
+deps =
+    withreq,pylint,mypy,pyright: -r{tox_root}/requirements-devel.txt
+    noreq: PyNaCl>=1.5.0
+    noreq,codespell: python-apt@git+https://salsa.debian.org/apt-team/python-apt.git@2.0.0
+extras =
+    noreq: dev
+package = wheel
+
+[testenv:py{38,310}-{withreq,noreq}-unit]
+description = Run the unit tests
+commands =
+    pytest {tty:--color=yes} --cov-report=xml:results/coverage-{env_name}.xml --junit-xml=results/test-results-{env_name}.xml tests/unit {posargs}
+
+[testenv:py{38,310}-{withreq,noreq}-legacy]
+description = Run the legacy unit tests
+commands =
+    pytest {tty:--color=yes} --cov-report=xml:results/coverage-{env_name}.xml --junit-xml=results/test-results-{env_name}.xml tests/legacy/unit/ {posargs}
+
+[testenv:format-black]
+description = Autoformat with black
+labels = fix
+deps = black
+skip_install = true
+# Note: this does not include `snapcraft_legacy` as it contains several files that need reformatting.
+commands = black setup.py snapcraft tests
+
+[testenv:format-ruff]
+description = Autoformat with ruff
+labels = fix
+deps = ruff
+skip_install = true
+commands =
+    ruff --fix setup.py snapcraft tests tools
```

Comment:
is `--fix` safe here?

Reply:
The `format` environments here actually run the formatting and are for local use rather than CI use. You can either run `tox -e format-ruff` to do the ruff formatting or (preferred) run `tox -m fix` to run all the auto-fixing tools (including black, as done above).

The CI runs the `lint` label, which includes `black` with `--check` and `ruff` without `--fix`.

My hope is that since `black` and `ruff` are so fast, we can eventually incorporate these into a pre-commit hook (assuming everyone's on board with that).